### DISCUSSION TITLE
ToListAsync extension method for IAsyncEnumerable<T>

### DISCUSSION THREAD

**Discussion by beepboopbork** [dotnet/csharplang#3935] [General] [answered]
The `IAsyncEnumerable<T>` interface is working pretty well for many things in the projects I've been working on. Seeing that it is still pretty new, I wouldn't expect the interface or extensions to be as complete as `IEnumerable<T>`, but I find myself writing my own extensions to make `IAsyncEnumerable<T>` usage work a little more like `IEnumerable<T>` in some cases.

In a similar sense that `IEnumerable<T>` has `.ToList<T>()` to immediately evaluate into a list, does it make sense to add a method or extension method to the C# libraries for `IAsyncEnumerable<T>` that does a similar thing for those times the enumerable needs to be immediately evaluated?

I like to think others may find this useful if built into the C# library, especially since `IEnumerable<T>.ToList<T>()` already exists. It has become especially useful in unit tests, and other situations where an async enumerator is useful but isn't always needed: as a simple example - in a method that reads a file async, yielding each line - sometimes each line can be read and processed one by one by the consuming code (and stopped without having to read the entire file) but other times all the lines need to be read before processing can begin.

I have created and have been using the following extension method to suit my needs:

```csharp        
public static async Task<List<T>> ToListAsync<T>(this IAsyncEnumerable<T> items)
{
    var evaluatedItems = new List<T>();
    await foreach (var item in items)
        evaluatedItems.Add(item);
    return evaluatedItems;
}
```
Example on usage, with a method doing something like this:
```csharp
public async IAsyncEnumerable<string> ReadAsync(string path)
{
    await using (var fileStream = File.OpenRead(path))
    using (var streamReader = new StreamReader(fileStream))
    {
        yield return await streamReader.ReadLineAsync();
    }
}
```
it could be used either way (similar to what is already possible with `IEnumerable<T>`):
```csharp
public async Task DoStuffLineByLineAsync()
{
    await foreach(var line in ReadAsync("some file path"))
        ProcessLine(line); // process each line
}

public async Task DoStuffAtOnceAsync()
{
    List<string> allLines = await ReadAsync("some file path").ToListAsync();
        CalculateSomething(allLines); // needed all the lines before processing 
}
```
Any thoughts on adding this, and possibly other `IEnumerable<T>` methods/extensions to `IAsyncEnumerable<T>`?

**Comment by sharwell** [comment]
I would not be surprised to see an `AsyncEnumerable` class appear (possibly as a third-party library) to help with two cases:

1. Inputs that are `IAsyncEnumerable<T>`
2. Predicates that return `ValueTask<bool>` and/or transformation functions that return `ValueTask<TResult>`

**Comment by HaloFour** [comment, marked as answer]
I'd suggest opening a discussion at the [Runtime](https://github.com/dotnet/runtime/issues) repository as they handle API discussions.  This repository is more for evolution of the C# language itself.

**Reply by CyrusNajmabadi** [reply]
This is definitely a good question.  However, @HaloFour  is correct that it's out of scope for CSharpLang.  I would recommend taking this to the Runtime repo as they manage adding new APIs.

**Reply by beepboopbork** [reply]
apologies to all, I'll open a discussion over there. Thanks for pointing me in the right direction!

**Comment by YairHalberstadt** [comment]
Check out System.Linq.Async by ReactiveExtensions: https://www.nuget.org/packages/System.Linq.Async

**Comment by MgSam** [comment]
For whatever reason, the BCL team is dead-set against including *any* extension methods for IAsyncEnumerable, instead delegating it all to Rx (which of course has no where near the level of support as the BCL)

https://github.com/dotnet/runtime/issues/31580

**Reply by CyrusNajmabadi** [reply]
@stephentoub and others in that thread gave perfectly reasonable explanations on why this is.  It may not be what you want, but it's up to them to decide what level of in-box vs out-of-box balance they desire.  If you don't like that you can take it up with them.  It isn't part of any csharplang discussion.

**Reply by MgSam** [reply]
> If you don't like that you can take it up with them. It isn't part of any csharplang discussion.

I didn't say it was relevant to csharplang. I was answering the question with the only remotely official response to this topic I've seen and linking a relevant discussion; which both you and the others who answered the question were unaware of.

I do find it a little odd that a member of the C# language design committee thinks that libraries supporting the language features they just spent years designing and building is not something that you guys should care about... [Microsoft politics](https://www.extremetech.com/wp-content/uploads/2013/07/microsoft-org-chart.jpg) 🙄

**Reply by CyrusNajmabadi** [reply]
> I do find it a little odd that a member of the C# language design committee thinks that libraries supporting the language features they just spent years designing and building is not something that you guys should care about

I think the design and building of those APIs is under their purview.  And the reasons and rational they've given for the decisions here seem perfectly reasonable.  If you want them to change anything there, the discussion should be with them.

**Reply by 333fred** [reply]
They're the subject matter experts. We certainly care about the experience of using C# as a whole, including the library experience. However, while we spend all of our time designing the semantics of the language, we don't spend all of our time thinking about the library itself, or how users will interact with them. We trust the folks running the BCL to make the right decisions. If anything, that's the exact _opposite_ of that org chart :)

**Reply by CyrusNajmabadi** [reply]
> Microsoft politics 🙄

This is literally a case of respecting another team and allowing them to make decisions in the areas they understand the best and have the most context on.  It's avoiding any sort of political interference.

**Reply by MgSam** [reply]
@333fred I'm commenting on the fact that it apparently hasn't even occurred to you guys prior to now what the library situation around `IAsyncEnumerable` is or that you should have any input on it. (A feature that shipped one year ago...)

Features cannot be successful if designed in isolation. If LINQ had been shipped without BCL support, it would have been a failure; of this I have little doubt. The C# team has made the need for the partnership with the BCL team clear with a lot of features, past and present. I don't understand the seeming lack of cooperation with this feature. It was shipped, supported in a (very) small smattering of APIs in the BCL, and otherwise abandoned, AFAICT. In the run-up to the release of C# 8.0, the language team widely alluded to the BCL needing a LINQ-like library for it, apparently with no follow-through. I can tell you how that story ends already- probably as well as async iterators have turned out in JavaScript/TypeScript (hint: with no library support - I've never seen them used anywhere - and it's been 3 years since they shipped).

Anyway, I agree with Cyrus that the discussion about whether to include it in the BCL should happen in runtime, which is why I linked the relevant discussion.

**Reply by CyrusNajmabadi** [reply]
> @333fred I'm commenting on the fact that it apparently hasn't even occurred to you guys prior to now what the library situation around IAsyncEnumerable is or that you should have any input on it.

The runtime position seems acceptable to me.  They've made a call if the support needs to be in box or out of box and given justification for it. 

> Features cannot be successful if designed in isolation.

They are not designed in isolation.

>  I don't understand the seeming lack of cooperation with this feature.

Because you are inferring lack of cooperation.  That's only something *you* have come up with.

**Reply by CyrusNajmabadi** [reply]
> If LINQ had been shipped without BCL support, it would have been a failure; of this I have little doubt. 

Nuget wasn't a thing at the time.  Things have changed.  Neither team believes that both must be in lockstep on a feature for something to be successful.  Libraries can be released without specific language features for them.  And language features can be released without specific in-box APIs for them.

**Reply by stephentoub** [reply]
> I'm commenting on the fact that it apparently hasn't even occurred to you guys prior to now what the library situation around IAsyncEnumerable is or that you should have any input on it.

This was discussed extensively with @MadsTorgersen and in C# LDM meetings.  It even shows up in the original proposal for the C# 8 feature.

**Reply by MgSam** [reply]
> Because you are inferring lack of cooperation. That's only something _you_ have come up with.

Cyrus, if you had been cooperating with the runtime team, you (C# team) would have already had this discussion internally with the runtime team a year+ ago and known their position that they do not want to support IAsyncEnumerable in the BCL. You wouldn't need a part-time community member to tell you about the decision. I don't enjoy getting into these back-and-forths with you but let's call a spade a spade here.

**Reply by jnm2** [reply]
@MgSam What makes you think Cyrus in particular was part of that discussion at that time? There are some strong assumptions on your part needed to hold up what you're saying. I don't see the point of challenging him about this.

**Reply by CyrusNajmabadi** [reply]
@MgSam My position has continued to be: this is conversation to have with the runtime team.  Regardless of the decisions and discussions that have happened previously, that is the appropriate team to be involving to see if anything will change here.  That has been entirely the messaging i've been adding to this discussion (both to the OP, and you).  The runtime may continue to believe their stated reasoning so far (which i have no reason to believe they would change), or they may adjust based on the presence of this new feedback.  In either case, that is the best and most appropriate venue to converse on this.

**Comment by HaloFour** [comment]
That said, if it's a question of visibility and discoverability that said MS NuGet packages exist, which team would that fall upon?  Like say if a developer tried to do `iae.ToList` could VS/Roslyn suggest, "hey, looks like you're trying to use LINQ on an `IAsyncEnumerable`, would you like to automagickally add this package and import these namespaces?"

**Reply by CyrusNajmabadi** [reply]
Yes.  That's tracked under this issue: https://github.com/dotnet/roslyn/issues/8757

**Reply by beepboopbork** [reply]
@HaloFour - I think you nailed why I posted (on the wrong repo, sorry!). I didn't realize there was an official package due to Nuget not showing the [dotnetframework](https://www.nuget.org/profiles/dotnetframework) or [Microsoft](https://www.nuget.org/profiles/Microsoft) as an owner like all the other packages I've pulled in to get extra "official" features. I've learned something today at least 😄
